### PR TITLE
Add unlerp and relerp

### LINKED ISSUE #8: Linear Interpolation Additions: unlerp / relerp
These two functions occur so frequently in situations involving linear interpolation that I wished they were built-in. Especially `relerp`, which I cite on the GMC all the time for things like health and scroll bars, room-to-screen coordinate conversions, etc.

```gml
///@func unlerp(a, b, v)
///@param {Real} a The first value.
///@param {Real} b The second value.
///@param {Real} v The interpolated value.
///@return {Real}
///@desc Return the interpolation coefficient t such that lerp(a, b, t) equals v.
function unlerp(a, b, v) {
    return (v-a)/(b-a);
}
```

```gml
///@func relerp(x0, x1, v, y0, y1)
///@param {Real} x0 The old range's first value.
///@param {Real} x1 The old range's second value.
///@param {Real} v The value to re-interpolate.
///@param {Real} y0 The new range's first value.
///@param {Real} y1 The new range's second value.
///@return {Real}
///@desc Given v in the old range (x0, x1), return its equivalent value in the new range (y0, y1).
function relerp(x0, x1, v, y0, y1) {
    return y0+(y1-y0)*(v-x0)/(x1-x0);
}
```

### PR DESCRIPTION
**Pull request by dicksonlaw583**
Resolves #8.

This implements `unlerp` and `relerp` as functions under math utilities and adds the corresponding tests.

### CODE REVIEW

File: Community Toolbox/scripts/utils_Math/utils_Math.gml [hunk 1]
```diff
@@ -83,4 +83,34 @@ function approach_angle(_current, _target, _step = 1) {
...
+///@param {Real} newto The new range's second value.
+///@return {Real}
+///@desc Given oldvalue in the old range (oldfrom, oldto), return its equivalent value in the new range (newfrom, newto).
+function relerp(_oldfrom, _oldto, _oldvalue, _newfrom, _newto) {
+    if (_oldto == _oldfrom) {
+        throw "relerp() old range has zero width.";
+    }
+    return _newfrom+(_newto-_newfrom)*(_oldvalue-_oldfrom)/(_oldto-_oldfrom);
+}
+#endregion
```
Comment (Alphish) [inline Community Toolbox/scripts/utils_Math/utils_Math.gml:116]:
```suggestion
#region Linear interpolation

/// @func unlerp(from,to,value)
/// @desc Returns the interpolation amount such that lerp(from, to, amount) equals the given value.
/// @arg {Real} from        The first value.
/// @arg {Real} to          The second value.
/// @arg {Real} value       The interpolated value.
/// @returns {Real}
function unlerp(_from, _to, _value) {
    if (_from == _to) {
        throw "unlerp() range has zero width.";
    }
    return (_value - _from) / (_to - _from);
}

/// @func relerp(oldfrom,oldto,oldvalue,newfrom,newto)
/// @desc Given an interpolated value in the old range (oldfrom, oldto), returns its equivalent value in the new range (newfrom, newto).
/// @arg {Real} oldfrom     The old range's first value.
/// @arg {Real} oldto       The old range's second value.
/// @arg {Real} oldvalue    The value to re-interpolate.
/// @arg {Real} newfrom     The new range's first value.
/// @arg {Real} newto       The new range's second value.
/// @returns {Real}
function relerp(_oldfrom, _oldto, _oldvalue, _newfrom, _newto) {
    if (_oldto == _oldfrom) {
        throw "relerp() old range has zero width.";
    }
    var _amount = (_oldvalue - _oldfrom) / (_oldto - _oldfrom);
    return _newfrom + (_newto - _newfrom) * _amount;
}

#endregion
```
Made spacing and JSDocs more consistent.

File: Community Toolbox/scripts/tests_UnlerpTests/tests_UnlerpTests.gml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    
+    static should_unlerp_body_in_reverse = function() {
+        given_range_from(67);
+        given_range_to(35);
+        given_amount(43);
+        when_unlerp_calculated();
+        then_result().should_be(0.75);
+    };
+
```
Comment (Alphish) [inline Community Toolbox/scripts/tests_UnlerpTests/tests_UnlerpTests.gml:83]:
```suggestion
    
    static should_throw_given_zero_width_range = function() {
        given_range_from(35);
        given_range_to(35);
        given_amount(43);
        try {
            when_unlerp_calculated();
            assert_fail("Exception should be thrown for a zero-width range, but it wasn't.");
        } catch (e) {
            assert_pass();
        }
    }
    
```

Comment (Alphish) [reply Community Toolbox/scripts/tests_UnlerpTests/tests_UnlerpTests.gml:83]: Added `unlerp` test for zero-width range.

File: Community Toolbox/scripts/tests_RelerpTests/tests_RelerpTests.gml [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    static should_relerp_body_in_reverse = function() {
+        given_old_range_from(67);
+        given_old_range_to(35);
+        given_amount(43);
+        given_new_range_from(1000);
+        given_new_range_to(1100);
+        when_relerp_calculated();
+        then_result().should_be(1075);
+    };
+
```
Comment (Alphish) [inline Community Toolbox/scripts/tests_RelerpTests/tests_RelerpTests.gml:103]:
```suggestion
    
    static should_throw_given_zero_width_old_range = function() {
        given_old_range_from(35);
        given_old_range_to(35);
        given_amount(43);
        given_new_range_from(1000);
        given_new_range_to(1100);
        given_amount(43);
        try {
            when_unlerp_calculated();
            assert_fail("Exception should be thrown for a zero-width old range, but it wasn't.");
        } catch (e) {
            assert_pass();
        }
    }
    
    static should_relerp_given_zero_width_new_range = function() {
        given_old_range_from(35);
        given_old_range_to(67);
        given_amount(43);
        given_new_range_from(1000);
        given_new_range_to(1000);
        when_relerp_calculated();
        then_result().should_be(1000);
    }
    
```
Added `relerp` for zero-width old range and new range.

**Review by Alphish — COMMENTED** [review]
Suggested a few additional tests and making the JSDocs and spacing more consistent. These aside, it's good to go. ^^

**Comment by dicksonlaw583** [comment]
@Alphish:

I have accepted your suggestions. Please feel free to merge any time you are ready.

**Review by Alphish — APPROVED** [review]

**Pull request merged**